Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLCommon.py

```diff
@@ -0,0 +1,261 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+from mantid.simpleapi import CropToComponent, logger, mtd
+from mantid.api import WorkspaceGroup
+from enum import IntEnum
+from math import fabs
+from os import path
+
+
+class AcqMode(IntEnum):
+    '''
+    |        |  MONO  |  KINETIC  |  TOF  |  REVENT  |
+    | X Unit |  Empty |   Empty   | Lambda|    TOF   |
+    | X Size |    1   |    >1     |  >1   |    >1    |
+    | IsHist?|    n   |     n     |   y   |     y    |
+    '''
+    MONO = 1 # standard monochromatic SANS
+    KINETIC = 2 # kinetic monochromatic SANS
+    TOF = 3 # TOF SANS [D33 only] (equidistant (LTOF) or non-equidistant (VTOF))
+    REVENT = 4 # rebinned monochromatic event data
+
+
+# Empty run rumber to act as a placeholder where a sample measurement is missing.
+EMPTY_TOKEN = '000000'
+# Quite often some samples are not measured at the some distances.
+# Still, it is preferred to reduce them together in one table.
+# Below is an example of why the blanks are needed:
+#    | samples D1 | samples D2 | samples D3 | sample TR |
+#    |     R1     |     R2     |     R3     |    R4     |
+#    |     R10    |     0      |     R11    |    R12    |
+# Here the sample 2 is not measured at distance 2.
+# But in order to reduce it together in a vectorized way and apply the transmissions
+# we will need to inject a blank measurement in the place of 0, with the same number of frames
+# as the sample measurement (so 1 if mono, >1 if kinetic mono).
+
+
+def get_run_number(value):
+    """
+    Extracts the run number from the first run out of the string value of a
+    multiple file property of numors
+    """
+    return path.splitext(path.basename(value.split(',')[0].split('+')[0]))[0]
+
+
+def needs_processing(property_value, process_reduction_type):
+    """
+    Checks whether a given unary reduction needs processing or is already cached
+    in ADS with expected name.
+    @param property_value: the string value of the corresponding MultipleFile
+                           input property
+    @param process_reduction_type: the reduction_type of process
+    """
+    do_process = False
+    ws_name = ''
+    if property_value:
+        run_number = get_run_number(property_value)
+        ws_name = run_number + '_' + process_reduction_type
+        if mtd.doesExist(ws_name):
```

Comment:
No, this is not supposed to work if a critical setting, such as the normalization changes. Supporting it will make things far more complicated. When changing this, one must clear the workspaces and start over. I'll add a note about this in the documentation.